compiler/rustc_ty_utils/src/consts.rs

```diff
@@ -18,53 +16,33 @@ use crate::errors::{GenericConstantTooComplex, GenericConstantTooComplexSub};
 
 /// Destructures array, ADT or tuple constants into the constants
 /// of their fields.
-fn destructure_const<'tcx>(
+fn destructure_adt_const<'tcx>(
     tcx: TyCtxt<'tcx>,
     const_: ty::Const<'tcx>,
-) -> ty::DestructuredConst<'tcx> {
+) -> ty::DestructuredAdtConst<'tcx> {
     let ty::ConstKind::Value(cv) = const_.kind() else {
         bug!("cannot destructure constant {:?}", const_)
     };
-    let branches = cv.to_branch();
-
-    let (fields, variant) = match cv.ty.kind() {
-        ty::Array(inner_ty, _) | ty::Slice(inner_ty) => {
-            // construct the consts for the elements of the array/slice
-            let field_consts = branches
-                .iter()
-                .map(|b| ty::Const::new_value(tcx, b.to_value().valtree, *inner_ty))
-                .collect::<Vec<_>>();
-            debug!(?field_consts);
-
-            (field_consts, None)
-        }
-        ty::Adt(def, _) if def.variants().is_empty() => bug!("unreachable"),
-        ty::Adt(def, _) => {
-            let (variant_idx, field_consts) = if def.is_enum() {
-                let (head, rest) = branches.split_first().unwrap();
-                (VariantIdx::from_u32(head.to_leaf().to_u32()), rest)
-            } else {
-                (FIRST_VARIANT, branches)
-            };
-            debug!(?field_consts);
 
-            (field_consts.to_vec(), Some(variant_idx))
-        }
-        ty::Tuple(elem_tys) => {
-            let fields = iter::zip(*elem_tys, branches)
-                .map(|(elem_ty, elem_valtree)| {
-                    ty::Const::new_value(tcx, elem_valtree.to_value().valtree, elem_ty)
-                })
-                .collect::<Vec<_>>();
+    let fields = cv.to_branch();
 
-            (fields, None)
+    let (variant, fields) = match cv.ty.kind() {
+        ty::Adt(def, _) if def.variants().is_empty() => {
+            bug!("unreachable")
         }
-        _ => bug!("cannot destructure constant {:?}", const_),
+        ty::Adt(def, _) if def.is_enum() => {
+            let (head, rest) = fields.split_first().unwrap();
+            (Some(VariantIdx::from_u32(head.to_leaf().to_u32())), rest)
+        }
+        ty::Adt(_, _) => (Some(FIRST_VARIANT), fields),
+        _ => bug!("destructure_adt_const called on non-ADT type: {:?}", cv.ty),
     };
 
-    let fields = tcx.arena.alloc_from_iter(fields);
+    debug!(?fields);
+
+    let fields = tcx.arena.alloc_slice(fields);
```

Comment:
is this necessary? I would expect that if we're (indirectly) taking a `ValTree<'tcx>` then we should be able to get a `&'tcx [Const<'tcx>]` out of it without needing to re-allocate anything